consensus-encoding/src/decode.rs

```diff
@@ -0,0 +1,166 @@
+use super::{Decoder, ReadError};
+
+/// Defines the decoder used for decoding the consensus type implementing this trait.
+pub trait Decode: Sized {
+    /// The decoder used when decoding this type.
+    type Decoder: Decoder<Value=Self> + Default;
+
+    /// Consensus-decodes from std reader.
+    #[cfg(feature = "std")]
+    fn consensus_decode<R: std::io::BufRead + ?Sized>(reader: &mut R) -> Result<Self, ReadError<<Self::Decoder as Decoder>::Error>> {
+        push_decode::decode_sync_with::<Self::Decoder, _>(reader, Default::default())
+    }
```

Comment:
This can be on `io::BufRead` and not feature gated soon as we release `io`, right? Is that the plan?

Reply:
No, I want to remove dependency on IO when we have this ready. I re-implement it in terms of `bitcoin-io` in `bitcoin::consensus` to maintain compatibility in transition period.

`no_std` users can simply use the lgio bridge function in `push_decode` or write their own bridge (it's just a few LOC).